File: acme-admin/src/main/java/com/acme/admin/controller/security/RoleAdminController.java
package com.acme.admin.controller.security;

import com.acme.admin.dto.security.RoleDtos.*;
import com.acme.admin.service.security.RoleService;
import jakarta.validation.Valid;
import org.springframework.security.access.prepost.PreAuthorize;
import org.springframework.web.bind.annotation.*;

import java.util.*;

@RestController
@RequestMapping("/api/admin/roles")
public class RoleAdminController {
  private final RoleService roles;

  public RoleAdminController(RoleService roles) {
    this.roles = roles;
  }

  @GetMapping
  @PreAuthorize("hasAuthority('roles.read')")
  public List<RoleDto> list() {
    return roles.list();
  }

  @PostMapping
  @PreAuthorize("hasAuthority('roles.write')")
  public RoleDto create(@Valid @RequestBody CreateRoleRequest req) {
    return roles.create(req);
  }

  @PutMapping("/{id}")
  @PreAuthorize("hasAuthority('roles.write')")
  public RoleDto update(@PathVariable UUID id, @Valid @RequestBody UpdateRoleRequest req) {
    return roles.update(id, req);
  }
}

File: acme-admin-dotnet/Controllers/RoleAdminController.cs
using Acme.Admin.Api.DTO;
using Acme.Admin.Api.Security;
using Acme.Admin.Api.Services;
using Acme.Admin.Api.Validation;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace Acme.Admin.Api.Controllers;

[ApiController]
[Route("api/admin/roles")]
public sealed class RoleAdminController(RoleService service) : ControllerBase
{
    [HttpGet]
    [Authorize(Policy = PermissionPolicies.RolesRead)]
    public async Task<ActionResult<PagedResult<RoleDtos.RoleDto>>> List(
        [FromQuery] int page = Paging.DefaultPage,
        [FromQuery] int size = Paging.DefaultSize,
        CancellationToken ct = default)
    {
        var (p, s) = Paging.Normalize(page, size);
        return Ok(await service.ListAsync(p, s, ct));
    }

    [HttpPost]
    [Authorize(Policy = PermissionPolicies.RolesWrite)]
    public async Task<ActionResult<RoleDtos.RoleDto>> Create([FromBody] RoleDtos.CreateRoleRequest req, CancellationToken ct)
    {
        return Ok(await service.CreateAsync(req, ct));
    }

    [HttpPut("{id:guid}")]
    [Authorize(Policy = PermissionPolicies.RolesWrite)]
    public async Task<ActionResult<RoleDtos.RoleDto>> Update([FromRoute] Guid id, [FromBody] RoleDtos.UpdateRoleRequest req, CancellationToken ct)
    {
        return Ok(await service.UpdateAsync(id, req, ct));
    }
}
